viahtml/stats.py

```diff
@@ -0,0 +1,108 @@
+"""uWSGI -> New Relic metrics collection logic."""
+
+from collections import defaultdict
+from logging import getLogger
+
+import requests
+from requests import RequestException
+
+LOG = getLogger(__name__)
+
+
+# pylint: disable=missing-yield-doc,missing-yield-type-doc
+
+
+class UWSGINewRelicStatsGenerator:
+    """A callable which returns stats from uWSGI stats for New Relic."""
+
+    # Mappings from various parts of the JSON structure we get back from
+    # uWSGI stats end-point to some suitable metrics names for New Relic
+    ROOT_STATS = {
+        "load": "System/Load",
+        "listen_queue": "ListenQueue/Size",
+        "listen_queue_errors": "ListenQueue/Errors",
+        "signal_queue": "SignalQueue/Size",
+    }
+
+    WORKER_STATS = {
+        "avg_rt": "Worker/AverageResponseTime",
+        "requests": "Worker/Request/Total",
+        "exceptions": "Worker/Request/Failed",
+        "harakiri_count": "Worker/Killed",
+        "running_time": "Worker/UpTime",
+        "rss": "Worker/Memory/Resident",
+        "vsz": "Worker/Memory/Virtual",
+    }
+
+    SOCKET_STATS = {
+        "queue": "Socket/Queue/Size",
+        "max_queue": "Socket/Queue/Max",
+    }
+
+    def __init__(self, stats_endpoint):
+        """Create a callable metrics generator.
+
+        :param stats_endpoint: The uWSGI stats end-point to call
+        """
+        self._stats_endpoint = stats_endpoint
+
+    def __call__(self):
+        """Generate stat name value pairs."""
+
+        for stat_name, values in self._get_stats():
+            LOG.debug("Reporting stat %s = %s", stat_name, values)
+            yield f"Custom/{stat_name}", values
+
+    def _get_stats(self):
+        # General "I'm alive!" type stuff
+        yield "Alive/Metrics", 1
+
+        try:
+            raw_stats = requests.get(self._stats_endpoint).json()
+        except RequestException as err:
+            LOG.debug("Stats collection from uWSGI failed with error: %s", err)
+
+            yield "Alive/App", 0
+            return
+
+        yield "Alive/App", 1
+
+        # General metrics
+        for key, stat_name in self.ROOT_STATS.items():
+            yield stat_name, raw_stats[key]
+
+        # Socket metrics
+        yield from self._stats_from_items(raw_stats["sockets"], self.SOCKET_STATS)
```

Comment:
I've only seen one socket locally, but who knows? Probably best to be on the safe side and use the aggregation function